solutions/8-preserve-identity/index.ts

```diff
@@ -0,0 +1,127 @@
+export function createMembrane<T extends object>(target: T) {
+  const revokeFnsCache = new RevokeFnsCache<T>();
+  const proxyIdentityCache = new ProxyIdentityCache<T>();
+  const proxy = createRevocableProxy(target, revokeFnsCache, proxyIdentityCache);
+  return {
+    membrane: proxy,
+    revoke: () => {
+      revokeFnsCache.revokeAll();
+    },
+  };
+}
+
+function createRevocableProxy<T extends object>(
+  target: T,
+  revokeFnsCache: RevokeFnsCache<T>,
+  proxyIdentityCache: ProxyIdentityCache<T>
+): T {
+  function wrapper(target: any, direction: Direction): T {
+    // return target if it's a primitive value
+    if (isPrimitive(target)) return target;
+    // Return the proxy if it already exists.
+    if (proxyIdentityCache.get(target, direction)) return proxyIdentityCache.get(target, direction);
+    const flippedDirection = proxyIdentityCache.flipDirection(direction);
+
+    // Create a proxy object that will forward all of the traps to a single function.
+    const proxyHandler = new Proxy(
```

Comment:
One open question is if this forwarding proxy could ever leak, as it won't be cleaned up by the revoke
one option is to also make it revokable

Reply:
But given that it's the handler to Proxy.Revocable, I assume revoke cleans it up